File: ios/Sources/FlowbizOnsite/Event.swift
import Foundation

/// Typed event catalog — the single typed entry point of SPEC §5.
///
/// Compile-time typing replaces the web Yup layer: there is no runtime schema
/// validation and no field enrichment. Integrators build complete typed
/// payloads; the SDK serializes them verbatim (snake_case wire keys, optional
/// fields omitted when nil). All cases carry `Sendable` value types (SPEC §3).
public enum Event: Sendable, Equatable {

    /// `page.view` — `path` is resolved against `FlowbizConfig.baseUri`
    /// (spec §5) into `page.url` and the remembered `context.url`;
    /// `title` ships as `page.title`. Both optional.
    case pageView(path: String? = nil, title: String? = nil)

    /// `account.login`.
    case accountLogin(user: User)

    /// `account.sync`.
    case accountSync(user: User)

    /// `product.view`.
    case productView(product: Product)

    /// `cart.sync`.
    case cartSync(cart: Cart)

    /// `cart.add`.
    case addToCart(products: [CartItem])

    /// `cart.item.update`.
    case cartItemUpdate(cartId: String, productId: String, sku: String, quantity: Int)

    /// `cart.setpostalcode`.
    case cartSetPostalCode(cartId: String, postalCode: String)

    /// `cart.setcoupon`.
    case cartSetCoupon(cartId: String, coupon: String)

    /// `checkout.step`.
    case checkoutStep(checkout: Checkout)

    /// `order.complete`.
    case orderComplete(order: Order)

    /// `order.cancel` — at least one of `orderId` / `cartId` should be provided.
    case orderCancel(orderId: String?, cartId: String?)
}

File: android/sdk/src/main/kotlin/com/flowbiz/onsite/Event.kt
package com.flowbiz.onsite

/**
 * Typed event catalog — the single typed entry point of SPEC §5.
 *
 * Compile-time typing replaces the web Yup layer: there is no runtime schema
 * validation and no field enrichment. Integrators build complete typed
 * payloads; the SDK serializes them verbatim (snake_case wire keys, optional
 * fields omitted when null).
 */
sealed class Event {

    /** `page.view` — [screenName] also drives the synthetic `app://<screenName>` context URL. */
    data class PageView(val screenName: String? = null) : Event()

    /** `account.login`. */
    data class AccountLogin(val user: User) : Event()

    /** `account.sync`. */
    data class AccountSync(val user: User) : Event()

    /** `product.view`. */
    data class ProductView(val product: Product) : Event()

    /** `cart.sync`. */
    data class CartSync(val cart: Cart) : Event()

    /** `cart.add`. */
    data class AddToCart(val products: List<CartItem>) : Event()

    /** `cart.item.update`. */
    data class CartItemUpdate(
        val cartId: String,
        val productId: String,
        val sku: String,
        val quantity: Int,
    ) : Event()

    /** `cart.setpostalcode`. */
    data class CartSetPostalCode(
        val cartId: String,
        val postalCode: String,
    ) : Event()

    /** `cart.setcoupon`. */
    data class CartSetCoupon(
        val cartId: String,
        val coupon: String,
    ) : Event()

    /** `checkout.step`. */
    data class CheckoutStep(val checkout: Checkout) : Event()

    /** `order.complete`. */
    data class OrderComplete(val order: Order) : Event()

    /** `order.cancel` — at least one of [orderId] / [cartId] should be provided. */
    data class OrderCancel(
        val orderId: String? = null,
        val cartId: String? = null,
    ) : Event()
}
